stubs/python-decouple/decouple.pyi

```diff
@@ -0,0 +1,98 @@
+def strtobool(value: str | bool) -> bool: ...  # undocumented
+
+class UndefinedValueError(Exception): ...
+class Undefined: ...
+
+undefined: Undefined
+
+class Config:
+    repository: RepositoryEmpty
+    def __init__(self, repository: RepositoryEmpty) -> None: ...
+    @overload
+    def get(self, option: str, default: str = ...) -> str: ...
+    @overload
+    def get(self, option: str, cast: Callable[..., _T]) -> _T: ...
+    @overload
+    def get(self, option: str, default: str, cast: Callable[..., _T]) -> _T: ...
+    @overload
+    def __call__(self, option: str, default: str = ...) -> str: ...
+    @overload
+    def __call__(self, option: str, cast: Callable[..., _T]) -> _T: ...
+    @overload
+    def __call__(self, option: str, default: str, cast: Callable[..., _T]) -> _T: ...
+
+class RepositoryEmpty:  # undocumented
+    def __init__(self, source: str = ..., encoding: str = ...) -> None: ...
+    def __contains__(self, key: str) -> bool: ...
+    def __getitem__(self, key: str) -> str: ...
+
+class RepositoryIni(RepositoryEmpty):
+    SECTION: str
+    parser: ConfigParser
+    def __init__(self, source: str, encoding: str = ...) -> None: ...
+    def __contains__(self, key: str) -> bool: ...
+    def __getitem__(self, key: str) -> str: ...
+
+class RepositoryEnv(RepositoryEmpty):
+    data: dict[str, str]
+    def __init__(self, source: str, encoding: str = ...) -> None: ...
+    def __contains__(self, key: str) -> bool: ...
+    def __getitem__(self, key: str) -> str: ...
+
+class RepositorySecret(RepositoryEmpty):
+    data: dict[str, str]
+    def __init__(self, source: str = ...) -> None: ...
+    def __contains__(self, key: str) -> bool: ...
+    def __getitem__(self, key: str) -> str: ...
+
+class AutoConfig:
+    SUPPORTED: OrderedDict[str, RepositoryEmpty]
+    encoding: str
+    search_path: str | None
+    config: Config | None
+    def __init__(self, search_path: str | None = None) -> None: ...
+    @overload
+    def __call__(self, option: str, default: str = ...) -> str: ...
+    @overload
+    def __call__(self, option: str, cast: Callable[..., _T]) -> _T: ...
+    @overload
+    def __call__(self, option: str, default: str, cast: Callable[..., _T]) -> _T: ...
+
+config: AutoConfig
+
+class Csv(Generic[_T, _TCsv]):
+    cast: Callable[..., _T]
+    delimiter: str
+    strip: str
+    post_process: Callable[..., _TCsv]
+    def __call__(self, value: str) -> _TCsv: ...
+    def __init__(
+        self, cast: Callable[..., _T] = str, delimiter: str = ..., strip: str = ..., post_process: Callable[..., _TCsv] = list[_T]
+    ) -> None: ...
+
+class Choices(Generic[_T]):
+    flat: Sequence[_T]
+    cast: Callable[..., _T]
+    choices: Sequence[_T]
+    def __init__(
+        self,
+        flat: Sequence[_T] | None = None,
+        cast: Callable[..., _T] = ...,
+        choices: Sequence[_T | tuple[Incomplete, _T]] | None = None,
+    ) -> None: ...
+    def __call__(self, value: str) -> _T: ...
```

Comment:
```suggestion
    cast: Callable[..., _T]
    choices: Sequence[_T]
    @overload
    def __init__(
        self: Choices[str],
        flat: Sequence[str] | None = None,
        cast: Callable[[str], str] = ...,
        choices: Sequence[tuple[str, str]] | None = None,
    ) -> None: ...
    @overload
    def __init__(
        self: Choices[_T],
        flat: Sequence[_T],
        cast: Callable[[str], _T] = ...,
        choices: Sequence[tuple[_T, str]] | None = None,
    ) -> None: ...
    @overload
    def __init__(
        self: Choices[_T],
        flat: None,
        cast: Callable[[str], _T],
        choices: Sequence[tuple[_T, str]] | None = None,
    ) -> None: ...
    @overload
    def __init__(
        self: Choices[_T],
        flat: None = None,
        cast: Callable[[str], _T] = ...,
        *,
        choices: Sequence[tuple[_T, str]],
    ) -> None: ...
    def __call__(self, value: str) -> _T: ...
```
Same deal here, this takes some overloads to bind correctly and the parameters  of cast are given, and choices does need to be a sequence of 2 value tuples, the second value of which is always a string.

Technically you could make this generic in the input value type as well, but since it is meant for dealing with environment variables, it probably makes sense to stick with `str` as all the examples have as well.

Reply:
This looks great!  Thanks again for the suggestion.

I have one issue, it seems like when `cast` is used by itself mypy seems to get confused.

```python
from decouple import Choices, Csv
from typing_extensions import reveal_type

# error: Argument "cast" to "Choices" has incompatible type "type[int]"; expected "Callable[[str], str]"  [arg-type]
reveal_type(Choices(cast=int)(""))

reveal_type(Choices(flat=[1, 2])("")) # note: Revealed type is "builtins.int"
reveal_type(Choices(choices=[(1, "one"), (2, "two")])("")) #note: Revealed type is "builtins.int"
reveal_type(Choices(cast=int, flat=[1, 2])("")) #note: Revealed type is "builtins.int"
reveal_type(Choices(cast=int, choices=[(1, "one"), (2, "two")])("")) #note: Revealed type is "builtins.int"
reveal_type(Choices(flat=[1, 2], choices=[(1, "one"), (2, "two")])("")) #note: Revealed type is "builtins.int"
reveal_type(Choices(cast=int, flat=[1, 2], choices=[(1, "one"), (2, "two")])("")) #note: Revealed type is "builtins.int"
```

Reply:
This needs one additional overload to cover that case:
```python
    @overload
    def __init__(
        self: Choices[_T],
        flat: Sequence[_T] | None = None,
        *,
        cast: Callable[[str], _T],
        choices: Sequence[tuple[_T, str]] | None = None,
    ) -> None: ...
```

Reply:
generic overload matching gets complicated really quickly if you want to cover every possible case...

Reply:
> generic overload matching gets complicated really quickly if you want to cover every possible case on arguments other than the first one or two...

Yeah, I know what you mean. 

> I wish overloads would let you violate the requirement of positional optional parameters having to come after all the required ones, it would get rid of a lot of redundant overloads, that have to currently be used in order to express both the positional and keyword version of the overload.
> 
> But that would of course require some changes to the language, since the parser would have to ignore these violations specifically only on overloads, the alternative would be to have some typing specific sentinel value that means the parameter is required in order to match that overload, but the parser can still treat it like an optional one. But I am getting off course here.

TBH, this is how I thought it would work but the more I dug into making this type module, the more I realized that wasn't the case.  But, now I know and thanks for the background!